sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -751,6 +751,16 @@ object ConstantFolding extends Rule[LogicalPlan] {
 
       // Fold expressions that are foldable.
       case e if e.foldable => Literal.create(e.eval(EmptyRow), e.dataType)
+
+      // Use associative property for integral type
+      case e if e.isInstanceOf[BinaryArithmetic] && e.dataType.isInstanceOf[IntegralType]
+        => e match {
+        case Add(Add(a, b), c) if b.foldable && c.foldable => Add(a, Add(b, c))
```

Comment:
what about `a + 1 + b + 2`? I think we need a more general approach, like reordering the `Add` nodes to put all literals together.